Homework 22

marcellovi opened the pull request.

marcellovi committed:
Homework 21

Actions & Wait

marcellovi committed:
Homework 22

POM - Page Object Model

Berichicko commented on src/test/java/page/BasePage.java:
```diff
@@ -0,0 +1,88 @@
+
+public class BasePage {
+
+    WebDriver driver;
+    WebDriverWait wait;
+    Actions actions;
```
I recommend: make the variables protected, because protected access gives the subclass a chance to use the helper method or variable, while preventing a nonrelated class from trying to use it.

```
    protected Actions actions;
    protected WebDriverWait wait;
    protected WebDriver driver;
```

marcellovi replied:
Thank You! I will make the changes now that I am seeing all the messages.

Berichicko commented on src/test/java/page/HomePage.java:
```diff
@@ -0,0 +1,57 @@
+
+public class HomePage extends BasePage{
+
+    By avatarIcon = By.cssSelector("img.avatar");
+    By playlistEditField = By.xpath("//input[@data-testid=\"inline-playlist-name-input\"]");
```
You should use encapsulation of non-static and non-final fields.

If a variable is used only by other methods of that class, it should be declared as private.

I recommend:  make all the locators and WebElements private, because private keyword in java allows most restrictive access to variables and methods and offer strongest form of Encapsulation. 

For example

    ```
    private By shuffleBtnLocator = By.cssSelector(".btn-shuffle-all");
    private By firstSongLocator = By.cssSelector(".play");

```

marcellovi replied:
Thank You. For some reason all msgs were going to my Spam. I only noticed when I got the msg here on InteliJ. I am going to Fix and Upload and go through all the other branches as well. 
Thank you again!

Berichicko commented on src/test/java/BaseTest.java:
```diff
@@ -1,11 +1,146 @@
+    public void btnPlayNextSong(){
+        WebElement buttonPlayNext = this.driver.findElement(By.cssSelector("i[data-testid='play-next-btn']"));
+        buttonPlayNext.click();
+    }
+
+    public void imgSoundBarVisible(){
+        WebElement buttonPlayNext = this.driver.findElement(By.cssSelector("div[data-testid='sound-bar-play']"));
+        buttonPlayNext.isDisplayed();
+    }
+
+    public void rightClickPlaylistMenu(String playlist){
+        WebElement delPlaylist = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//section[@id=\"playlists\"]//a[contains(text(),\"" + playlist + "\")]")));
+        action.contextClick(delPlaylist).perform(); // Right click
+    }
+
+    public void selectSubMenuPlaylistOption(String option){
+        WebElement liDeleteTxt = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//nav//li[contains(text(),\""+ option +"\")]")));
+        liDeleteTxt.click();
+    }
+
+    public void updatePlaylistTextBoxContent(String playlistName) {
+        WebElement txtEditPlaylist = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//input[@data-testid=\"inline-playlist-name-input\"]")));
+        //txtEditPlaylist.clear(); It doesnt work - when I use clear right after the txtbox will become hidden
+        txtEditPlaylist.sendKeys(Keys.chord(Keys.CONTROL,"a"),playlistName);
+        txtEditPlaylist.sendKeys(Keys.ENTER);
+    }
+
+    // creates playlist
+    public void createPlaylist(String playlistName)  {
+
+        WebElement btnPlusIcon = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("i[data-testid=\"sidebar-create-playlist-btn\"]")));
+        btnPlusIcon.click();
+
+        WebElement liAddNewPlaylist = wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("li[data-testid=\"playlist-context-menu-create-simple\"")));
+        liAddNewPlaylist.click();
+
+        WebElement txtPlaylist = wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//input[@name=\"name\"]")));
+        txtPlaylist.sendKeys(playlistName);
+        txtPlaylist.sendKeys(Keys.ENTER);
+    }
```
Page Object Model (POM): recommended to associate the methods with the corresponding page class where they are relevant. In the POM, each page or section of a web application is represented by a separate class. These classes encapsulate the elements and actions specific to that page or section.

Move these methods to the corresponding class according to the POM.

For example
![Screenshot_2](https://github.com/YuriiLyndiuk/A57-QA-Automation-Framework/assets/104761482/b07cd7c1-7977-4e53-80ea-8045ac17f779)

Berichicko commented on src/test/java/BaseTest.java:
```diff
@@ -1,11 +1,146 @@
+    @Parameters({"baseUrl"})
+    public void initMethod(String baseUrl){
+        options = new ChromeOptions();
+        options.addArguments("--remote-allow-origins=*");
```
Add multiple arguments:
`--disable-notifications`: disabling notifications
`--remote-allow-origins=*`: allowing remote origins
`--incognito`: launching in incognito mode
`--start-maximized`: launching in maximized mode
`"excludeSwitches", new String[]{"enable-automation"} `: Disable notification Chrome is being controlled by automated test software in ChromeDriver
**These settings are enabled to simplify and improve the process of automated testing.**
_For example_
![image](https://github.com/Test-Pro-Bootcamp/A57-QA-Automation-Framework/assets/104761482/b05c1900-365a-491c-a399-92b5b3bdc666)
```
                WebDriverManager.chromedriver().setup();
                ChromeOptions options = new ChromeOptions();
                options.addArguments("--disable-notifications","--remote-allow-origins=*", "--incognito","--start-maximized");
                options.setExperimentalOption("excludeSwitches", new String[]{"enable-automation"});
                return new ChromeDriver(options);

```

marcellovi committed:
Suggested Changes

- Make BasePage variables protected so only Subclasses can access it
- Add multiple arguments for ChromeOptions
- Respect encapsulation by declaring variables as private.
